File: legacy /src/osv.ts
const OSV_QUERY_URL = 'https://api.osv.dev/v1/query';

export type OSVSeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW' | 'UNKNOWN';

export interface OSVAdvisory {
  id: string;
  summary?: string;
  severity: OSVSeverity;
}

export interface OSVResult {
  count: number;
  maxSeverity: OSVSeverity | null;
  advisories: OSVAdvisory[];
}

interface OSVVuln {
  id: string;
  summary?: string;
  severity?: Array<{ type: string; score: string }>;
  database_specific?: { severity?: string };
}

interface OSVQueryResponse {
  vulns?: OSVVuln[];
}

function cvssScoreToSeverity(score: string): OSVSeverity {
  // Extract base score from CVSS vector string (e.g. "CVSS:3.1/.../AV:N/.../")
  // or plain numeric score
  const numeric = parseFloat(score);
  if (!isNaN(numeric)) {
    if (numeric >= 9.0) return 'CRITICAL';
    if (numeric >= 7.0) return 'HIGH';
    if (numeric >= 4.0) return 'MEDIUM';
    if (numeric >= 0.1) return 'LOW';
    return 'UNKNOWN';
  }
  // CVSS vector string — extract the score differently; treat as UNKNOWN
  return 'UNKNOWN';
}

function parseSeverity(vuln: OSVVuln): OSVSeverity {
  // Prefer explicit database_specific severity string
  const dbSev = vuln.database_specific?.severity?.toUpperCase();
  if (dbSev === 'CRITICAL' || dbSev === 'HIGH' || dbSev === 'MEDIUM' || dbSev === 'LOW') {
    return dbSev;
  }

  // Fall back to CVSS score parsing
  for (const s of vuln.severity ?? []) {
    if (s.type === 'CVSS_V3' || s.type === 'CVSS_V2') {
      return cvssScoreToSeverity(s.score);
    }
  }

  return 'UNKNOWN';
}

const SEVERITY_RANK: Record<OSVSeverity, number> = {
  CRITICAL: 4,
  HIGH: 3,
  MEDIUM: 2,
  LOW: 1,
  UNKNOWN: 0,
};

function maxSeverity(severities: OSVSeverity[]): OSVSeverity | null {
  if (severities.length === 0) return null;
  return severities.reduce((a, b) => (SEVERITY_RANK[a] >= SEVERITY_RANK[b] ? a : b));
}

/**
 * Query the OSV database for known advisories affecting a GitHub repo.
 * Uses the pkg:github PURL which covers GitHub Actions and similar ecosystems.
 * Returns null on any error or timeout — never blocks installation.
 */
export async function fetchOSVAdvisories(
  ownerRepo: string,
  timeoutMs = 3000
): Promise<OSVResult | null> {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);

    const response = await fetch(OSV_QUERY_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        package: { purl: `pkg:github/${ownerRepo}` },
      }),
      signal: controller.signal,
    });
    clearTimeout(timeout);

    if (!response.ok) return null;

    const data = (await response.json()) as OSVQueryResponse;
    const vulns = data.vulns ?? [];

    const advisories: OSVAdvisory[] = vulns.map((v) => ({
      id: v.id,
      summary: v.summary,
      severity: parseSeverity(v),
    }));

    return {
      count: advisories.length,
      maxSeverity: maxSeverity(advisories.map((a) => a.severity)),
      advisories,
    };
  } catch {
    return null;
  }
}

File: src/osv.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type OSVSeverity string

const (
	OSVCritical OSVSeverity = "CRITICAL"
	OSVHigh     OSVSeverity = "HIGH"
	OSVMedium   OSVSeverity = "MEDIUM"
	OSVLow      OSVSeverity = "LOW"
	OSVUnknown  OSVSeverity = "UNKNOWN"
)

type OSVAdvisory struct {
	ID       string
	Summary  string
	Severity OSVSeverity
}

type OSVResult struct {
	Count       int
	MaxSeverity OSVSeverity
	Advisories  []OSVAdvisory
}

var osvQueryURL = "https://api.osv.dev/v1/query"

var severityRank = map[OSVSeverity]int{
	OSVCritical: 4,
	OSVHigh:     3,
	OSVMedium:   2,
	OSVLow:      1,
	OSVUnknown:  0,
}

func maxOSVSeverity(sevs []OSVSeverity) OSVSeverity {
	if len(sevs) == 0 {
		return OSVUnknown
	}
	max := sevs[0]
	for _, s := range sevs[1:] {
		if severityRank[s] > severityRank[max] {
			max = s
		}
	}
	return max
}

func cvssScoreToSeverity(score string) OSVSeverity {
	var f float64
	if _, err := parseFloat(score, &f); err == nil {
		switch {
		case f >= 9.0:
			return OSVCritical
		case f >= 7.0:
			return OSVHigh
		case f >= 4.0:
			return OSVMedium
		case f >= 0.1:
			return OSVLow
		}
	}
	return OSVUnknown
}

func parseFloat(s string, f *float64) (interface{}, error) {
	// Simple float parse
	var result float64
	n, err := fmt.Sscanf(s, "%f", &result)
	if n == 1 && err == nil {
		*f = result
		return nil, nil
	}
	return nil, fmt.Errorf("not a float")
}

func parseOSVSeverity(vuln map[string]interface{}) OSVSeverity {
	// Check database_specific.severity
	if dbSpec, ok := vuln["database_specific"].(map[string]interface{}); ok {
		if sev, ok := dbSpec["severity"].(string); ok {
			upper := strings.ToUpper(sev)
			switch OSVSeverity(upper) {
			case OSVCritical, OSVHigh, OSVMedium, OSVLow:
				return OSVSeverity(upper)
			}
		}
	}

	// Fall back to CVSS
	if sevArr, ok := vuln["severity"].([]interface{}); ok {
		for _, s := range sevArr {
			if sMap, ok := s.(map[string]interface{}); ok {
				t, _ := sMap["type"].(string)
				score, _ := sMap["score"].(string)
				if (t == "CVSS_V3" || t == "CVSS_V2") && score != "" {
					return cvssScoreToSeverity(score)
				}
			}
		}
	}
	return OSVUnknown
}

func fetchOSVAdvisories(ownerRepo string, timeoutMs int) *OSVResult {
	if timeoutMs <= 0 {
		timeoutMs = 3000
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutMs)*time.Millisecond)
	defer cancel()

	payload := map[string]interface{}{
		"package": map[string]string{
			"purl": "pkg:github/" + ownerRepo,
		},
	}
	body, _ := json.Marshal(payload)

	req, err := http.NewRequestWithContext(ctx, "POST", osvQueryURL, bytes.NewReader(body))
	if err != nil {
		return nil
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil
	}

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil
	}

	var result struct {
		Vulns []map[string]interface{} `json:"vulns"`
	}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil
	}

	var advisories []OSVAdvisory
	for _, v := range result.Vulns {
		id, _ := v["id"].(string)
		summary, _ := v["summary"].(string)
		sev := parseOSVSeverity(v)
		advisories = append(advisories, OSVAdvisory{ID: id, Summary: summary, Severity: sev})
	}

	if len(advisories) == 0 {
		return &OSVResult{Count: 0}
	}

	sevs := make([]OSVSeverity, len(advisories))
	for i, a := range advisories {
		sevs[i] = a.Severity
	}

	return &OSVResult{
		Count:       len(advisories),
		MaxSeverity: maxOSVSeverity(sevs),
		Advisories:  advisories,
	}
}

// Silence unused imports
var _ = fmt.Sprintf
